### PR TITLE
Restore "sort user-visible resources, objectives, vocab and introduced code alphabetically"

### PR DESCRIPTION
**Pull request by davidsbailey**
Reverts code-dot-org/code-dot-org#43090, restoring https://github.com/code-dot-org/code-dot-org/pull/42981. 

The original PR broke several lesson plans and rollup pages (see examples below). here is one honeybadger error:   https://app.honeybadger.io/projects/3240/faults/82042029

Summary of how we protect against this kind of problem in this PR:
* new default value for `programming_expression.syntax` and `objective.description` added in this PR
* new presence requirement on `resource.name` added in this PR. i also verified that there are no resources on levelbuilder with blank name
* `vocabulary` and `standard` are sorted on fields which already have "not null" db constraints

## Testing story

Manually verified that this fixes the following pages:
* http://localhost-studio.code.org:3000/s/coursef-2021/lessons/5
* http://localhost-studio.code.org:3000/s/coursed-2021/vocab

also verified that the new unit tests does not pass without the code in this PR

### CODE REVIEW

File: dashboard/app/models/resource.rb [hunk 1]
```diff
@@ -34,6 +34,8 @@ class Resource < ApplicationRecord
   KEY_RE = /\A#{KEY_CHAR_RE}+\Z/
   validates_format_of :key, with: KEY_RE, message: "must contain only lowercase alphanumeric characters, dashes, and underscores; got \"%{value}\"."
 
+  validates_presence_of :name
```
Comment (bethanyaconnor) [inline dashboard/app/models/resource.rb:37]: While it's not relevant to this PR, I think `url` should also be required here. I can do it in a separate PR though if you'd prefer to keep this one cleaner.

**Review by bethanyaconnor — APPROVED** [review]
Objectives really shouldn't have a nil description. Do we have a task to track? I'm not sure how that would look on the lesson plan.

I've noticed the nil syntax on programming expressions before and it does create some weirdness in the edit experience as well. I can look into that as part of the code docs work I'm doing now.

File: dashboard/app/models/resource.rb (hunk 1 above)
Comment (davidsbailey) [reply dashboard/app/models/resource.rb:37]: 100% agree, fortunately url is already required in the db schema: https://github.com/code-dot-org/code-dot-org/blob/08f5916adf6988088b29f9167fc8c527c42569a6/dashboard/app/models/resource.rb#L7

**Comment by dmcavoy** [comment]
> Objectives really shouldn't have a nil description. Do we have a task to track? I'm not sure how that would look on the lesson plan.

Here is the task to track https://codedotorg.atlassian.net/browse/PLAT-1390

**Review by dmcavoy — APPROVED** [review]
I like the new test. That is a nice way to deal with some of what we were talking about yesterday in standup

**Comment by davidsbailey** [comment]
> Objectives really shouldn't have a nil description. Do we have a task to track? I'm not sure how that would look on the lesson plan.

FWIW, it looks like `validates_presence_of` can be used on the description (and `syntax` below) even though it is a serialized attribute.

> I've noticed the nil syntax on programming expressions before and it does create some weirdness in the edit experience as well. I can look into that as part of the code docs work I'm doing now.

Thank you for picking that up!

**Pull request merged**